src/main/java/net/neoforged/neoforge/common/crafting/NBTIngredient.java

```diff
@@ -68,33 +89,64 @@ public boolean synchronizeWithContents() {
     }
 
     private static boolean compareStacksWithNBT(ItemStack left, ItemStack right) {
-        return left.getItem() == right.getItem() && NbtUtils.compareNbt(left.getTag(), right.getTag(), true);
+        return left.is(right.getItem()) && NbtUtils.compareNbt(left.getTag(), right.getTag(), true) && NbtUtils.compareNbt(left.serializeAttachments(), right.serializeAttachments(), true);
     }
 
     /**
      * Creates a new ingredient matching any item from the list, containing the given NBT
      */
     public static NBTIngredient of(boolean strict, CompoundTag nbt, ItemLike... items) {
-        return new NBTIngredient(Arrays.stream(items).map(ItemLike::asItem).collect(Collectors.toSet()), nbt, strict);
+        Objects.requireNonNull(nbt, "NBT Ingredient requires NBT");
+        return new NBTIngredient(Arrays.stream(items).map(ItemLike::asItem).collect(Collectors.toSet()), nbt, null, strict);
+    }
+
+    /**
+     * Creates a new ingredient matching any item from the list, containing the given NBT
+     */
+    public static NBTIngredient of(boolean strict, CompoundTag nbt, CompoundTag attachmentNbt, ItemLike... items) {
+        Objects.requireNonNull(nbt, "NBT Ingredient requires NBT");
+        Objects.requireNonNull(attachmentNbt, "NBT Ingredient requires attachment NBT");
```

Comment:
In this case I would allow passing `@Nullable`. This could be used from the `of(boolean, ItemStack)` factory as well.